frame/composable-traits/src/defi.rs

```diff
@@ -70,15 +73,34 @@ pub struct CurrencyPair<AssetId> {
 }
 
 impl<AssetId: PartialEq> CurrencyPair<AssetId> {
-	pub fn is_valid(&self) -> bool {
-		self.base != self.quote
+	pub fn new(base: AssetId, quote: AssetId) -> Self {
+		Self { base, quote }
+	}
+
+	///```rust
+	/// let pair = composable_traits::defi::CurrencyPair::<u128>::new(13, 42);
+	/// let slice =  pair.as_slice();
+	/// assert_eq!(slice[0], pair.base);
+	/// assert_eq!(slice[1], pair.quote);
+	/// drop(pair);
+	/// // next line will fail to compile
+	/// // let _ = slice[0];
+	/// ```
```

Comment:
You can use `compile_fail` to add a doctest that should fail to compile, so that it it still checked.